server/handler/auth.go

```diff
@@ -34,21 +35,30 @@ func OAuthCallback(
 		}
 
 		code := r.FormValue("code")
-		user, err := oAuth.GetUser(r.Context(), code)
+		ghUser, err := oAuth.GetUser(r.Context(), code)
 		if err != nil {
 			logger.Errorf("oauth get user error: %s", err)
 			// FIXME can it be not server error? for wrong code
 			write(w, r, serializer.NewEmptyResponse(), err)
 			return
 		}
 
-		// FIXME with real repo we need to check does user exists already or not
-		if err := userRepo.Create(user); err != nil {
-			logger.Errorf("can't create user: %s", err)
+		user, err := userRepo.Get(ghUser.Login)
+		if err != nil {
+			logger.Error(err)
 			write(w, r, serializer.NewEmptyResponse(), err)
 			return
 		}
 
+		if user == nil {
+			user, err = userRepo.Create(ghUser.Login, ghUser.Username, ghUser.AvatarURL, model.Requester)
```

Comment:
I noticed this pattern with passing fields of structs a few times already.
No changes are required now, but I prefer to pass a model to a repository.
It makes life easier when the model starts growing.

Reply:
Makes sense. But the problem, in this specific case, is that there is a field of the User model that is not up to you, the ID will be chosen by the DB.

You could create a User model and leave the ID as 0, but that can lead to bugs in my opinion... you may be tempted to continue using the `User` object you created. In the current code it is clear for the caller that the complete `User` object will be provided by `Create`.

Any other opinions?

Reply:
Sorry I don't get the problem with ID.

```
u := User{login,name,...}
err := repo.save(&u)
fmt.Println(u.ID) // prints user id
```

keep id as zero value for objects that aren't saved yet is totally normal.

You can see in for example in [most popular ORM for go](https://github.com/jinzhu/gorm)

Reply:
done in b09114c